### PR TITLE
sequence

### PR DESCRIPTION
**Pull request by malud**
Determine a possible sequence of roundtrips due to variable references.

Additionally there are some internal improvements access the backend response variables for all layers like endpoint, custom logs and error handling.

---
<details>
    <summary>Reviewer checklist</summary>
    <ul>
        <li>Read PR description: a summary about the changes is required</li>
        <li>Changelog updated</li>
        <li>Documentation: docs/{Reference, Cli, ...}, Docker and cli help/usage</li>
        <li>Pulled branch, manually tested</li>
        <li>Verified requirements are met</li>
        <li>Reviewed the code</li>
        <li>Reviewed the related tests</li>
    </ul>
</details>

### CODE REVIEW

File: server/http_endpoints_test.go [hunk 1]
```diff
@@ -437,3 +441,178 @@ func TestHTTPServer_NoGzipForSmallContent(t *testing.T) {
...
+
+	logs := hook.AllEntries()
+
+	var ctxCanceledSeen, statusOKseen bool
+	for _, entry := range logs {
+		if entry.Data["type"] != "couper_backend" {
+			continue
+		}
+
+		if strings.Contains(entry.Message, context.Canceled.Error()) {
```
Comment (johakoch) [inline server/http_endpoints_test.go:551]:
```suggestion
		path, _ := entry.Data["request"].(logging.Fields)["path"]

		if strings.Contains(entry.Message, context.Canceled.Error()) {
```

File: server/http_endpoints_test.go [hunk 2]
```diff
@@ -437,3 +441,178 @@ func TestHTTPServer_NoGzipForSmallContent(t *testing.T) {
...
+			ctxCanceledSeen = true
+		}
+
+		if entry.Message == "" && entry.Data["status"] == 200 {
+			statusOKseen = true
+		}
```
Comment (johakoch) [inline server/http_endpoints_test.go:557]:
```suggestion
		path, _ := entry.Data["request"].(logging.Fields)["path"]

		if strings.Contains(entry.Message, context.Canceled.Error()) {
			ctxCanceledSeen = true
			if path != "/" {
				t.Errorf("expected '/' to fail")
			}
		}

		if entry.Message == "" && entry.Data["status"] == 200 {
			statusOKseen = true
			if path != "/reflect" {
				t.Errorf("expected '/reflect' to be ok")
			}
		}
```

File: server/http_endpoints_test.go [hunk 3]
```diff
@@ -437,3 +441,178 @@ func TestHTTPServer_NoGzipForSmallContent(t *testing.T) {
...
+			continue
+		}
+
+		if entry.Message == "backend timeout error: deadline exceeded" {
+			ctxDeadlineSeen = true
+		}
+
+		if entry.Message == "" && entry.Data["status"] == 200 {
+			statusOKseen = true
+		}
```
Comment (johakoch) [inline server/http_endpoints_test.go:611]:
```suggestion
		path, _ := entry.Data["request"].(logging.Fields)["path"]

		if entry.Message == "backend timeout error: deadline exceeded" {
			ctxDeadlineSeen = true
			if path != "/" {
				t.Errorf("expected '/' to fail")
			}
		}

		if entry.Message == "" && entry.Data["status"] == 200 {
			statusOKseen = true
			if path != "/reflect" {
				t.Errorf("expected '/reflect' to be ok")
			}
		}
```

Comment (johakoch) [reply server/http_endpoints_test.go:611]: Needs
```
	"github.com/avenga/couper/logging"
```

File: server/http_endpoints_test.go [hunk 4]
```diff
@@ -5,15 +5,19 @@ import (
...
 	"path"
 	"reflect"
 	"strconv"
 	"strings"
 	"testing"
 	"time"
 
+	"github.com/google/go-cmp/cmp"
+
 	"github.com/avenga/couper/internal/test"
```
Comment (malud) [inline server/http_endpoints_test.go:21]:
```suggestion
	"github.com/avenga/couper/internal/test"
	"github.com/avenga/couper/logging"
```

File: config/runtime/endpoint.go [hunk 5]
```diff
@@ -145,25 +160,158 @@ func newEndpointOptions(confCtx *hcl.EvalContext, endpointConf *config.Endpoint,
...
+		Requests:      requests,
+		Sequences:     sequences,
+		Response:      response,
+		ServerOpts:    serverOptions,
 	}, nil
 }
+
+// newSequence lookups any request related dependency and sort them into a sequence.
+// Also return left-overs for parallel usage.
+func newSequence(proxies map[string]*producer.Proxy, requests map[string]*producer.Request, errCh chan<- error,
```
Comment (johakoch) [inline config/runtime/endpoint.go:186]: rename to `newSequences()`, as it creates `producer.Sequences` (plural)

File: docs/REFERENCE.md [hunk 6]
```diff
@@ -142,37 +142,41 @@ The `proxy` block creates and executes a proxy request to a backend service.
...
+| `backend`               | string       | -       | [Backend Block](#backend-block) reference, defined in [Definitions Block](#definitions-block)                                                                                                                                                                                                    | &#9888; required, if no [Backend Block](#backend-block) or `url` attribute is defined. | `backend = "foo"`   |
+| `expected_status`       | list<number> | -       | If defined, the [Backend Block](#backend-block) response status code will be verified against this list of codes. If the status-code is unexpected an [`unexpected_status` error](./ERRORS.md#error-types) can be handled with an [`error_handler`](./ERRORS.md#endpoint-related-error_handler). | -                                                                                      | -                   |
```
Comment (johakoch) [inline docs/REFERENCE.md:152]:
```suggestion
| `expected_status`       | list<number> | -       | If defined, the response status code will be verified against this list of codes. If the status-code is unexpected an [`unexpected_status` error](./ERRORS.md#error-types) can be handled with an [`error_handler`](./ERRORS.md#endpoint-related-error_handler). | -                                                                                      | -                   |
```

File: docs/REFERENCE.md [hunk 7]
```diff
@@ -142,37 +142,41 @@ The `proxy` block creates and executes a proxy request to a backend service.
...
+| `body`            | string                                | -       | -                                                                                                                                                                                                                                                                                                | Creates implicit default `Content-Type: text/plain` header field.                                                                                                                      | -                 |
+| `expected_status` | list<number>                          | -       | If defined, the [Backend Block](#backend-block) response status code will be verified against this list of codes. If the status-code is unexpected an [`unexpected_status` error](./ERRORS.md#error-types) can be handled with an [`error_handler`](./ERRORS.md#endpoint-related-error_handler). | -                                                                                                                                                                                      | -                 |
```
Comment (johakoch) [inline docs/REFERENCE.md:173]:
```suggestion
| `expected_status` | list<number>                          | -       | If defined, the response status code will be verified against this list of codes. If the status-code is unexpected an [`unexpected_status` error](./ERRORS.md#error-types) can be handled with an [`error_handler`](./ERRORS.md#endpoint-related-error_handler). | -                                                                                                                                                                                      | -                 |
```

File: handler/transport/backend_test.go [hunk 8]
```diff
@@ -274,6 +274,7 @@ func TestBackend_RoundTrip_Validation(t *testing.T) {
 }
 
 func TestBackend_director(t *testing.T) {
+	t.Skip("TODO: test-hook to obtain values")
```
Comment (johakoch) [inline handler/transport/backend_test.go:277]: Needs resolve

File: config/runtime/server.go [hunk 9]
```diff
@@ -386,18 +345,12 @@ func NewServerConfiguration(conf *config.Couper, log *logrus.Entry, memStore *ca
...
-				bodies = append(bodies, apiBodies...)
-			}
-			epBodies := bodiesWithACBodies(
-				conf.Definitions, endpointConf.Remain,
-				endpointConf.AccessControl, endpointConf.DisableAccessControl,
-			)
+				apiBodies := bodiesWithACBodies(conf.Definitions, parentAPI.AccessControl, parentAPI.DisableAccessControl)
+				bodies = append(bodies, append(apiBodies, parentAPI.HCLBody())...)
+			}
+			acBodies := bodiesWithACBodies(conf.Definitions, endpointConf.AccessControl, endpointConf.DisableAccessControl)
```
Comment (johakoch) [inline config/runtime/server.go:346]: Why is this renamed to `acBodies`?

File: config/runtime/server.go [hunk 10]
```diff
@@ -180,10 +147,8 @@ func NewServerConfiguration(conf *config.Couper, log *logrus.Entry, memStore *ca
 			}
 		}
 
-		serverBodies := bodiesWithACBodies(
-			conf.Definitions, srvConf.Remain,
-			srvConf.AccessControl, srvConf.DisableAccessControl,
-		)
+		serverBodies := bodiesWithACBodies(conf.Definitions, srvConf.AccessControl, srvConf.DisableAccessControl)
+		serverBodies = append(serverBodies, srvConf.HCLBody())
```
Comment (johakoch) [inline config/runtime/server.go:146]: Why `srvConf.HCLBody()` instead of `srvConf.Remain`, etc.?

File: docs/ERRORS.md [hunk 11]
```diff
@@ -29,7 +33,11 @@ Since an [AC](#access-control-error_handler) error handler is rather general tha
 same access-control. Reacting to those more path specific errors you can define one or more `error_handler` blocks within
 an `api` or `endpoint` block.
 
-### `error_handler` specification
+## Endpoint related `error_handler`
+
+TBA
```
Comment (johakoch) [inline docs/ERRORS.md:38]: docu missing yet

File: config/configload/load.go [hunk 12]
```diff
@@ -2,25 +2,20 @@ package configload
...
 	"github.com/avenga/couper/config"
-	hclbody "github.com/avenga/couper/config/body"
 	"github.com/avenga/couper/config/configload/collect"
 	"github.com/avenga/couper/config/parser"
 	"github.com/avenga/couper/config/reader"
 	"github.com/avenga/couper/errors"
 	"github.com/avenga/couper/eval"
 	"github.com/avenga/couper/eval/lib"
 	"github.com/avenga/couper/internal/seetie"
+	"github.com/hashicorp/hcl/v2"
```
Comment (johakoch) [inline config/configload/load.go:17]: fix order of imports:
```
	"github.com/hashicorp/hcl/v2"
	"github.com/hashicorp/hcl/v2/gohcl"

	"github.com/avenga/couper/config"
	...
```

**Review by johakoch — CHANGES_REQUESTED** [review]

**Review by johakoch — DISMISSED** [review]

**Review by johakoch — APPROVED** [review]

**Pull request merged**